### PR TITLE
[ra-core] use react-query useQuery for permissions fetching

### PR DESCRIPTION
**Pull request by hiaselhans**
I wanted to hook up the builtin permission system from react-admin and I experienced a ton of queries: one per permission check.
the docs have a hidden (pro/paid flagged) hint on how to solve this:
https://marmelab.com/react-admin/AuthRBAC.html#performance

since we are using react-query already all-over why don't we use it for the permission-fetching?
This adds a lot of sane defaults: re-fetch on window focus, cache invalidation on logout, retries with back-off, ...

I think the same could be applied to `useCheckAuth`, `useGetIdentity`,... but I'd like to hear your comments first.

### CODE REVIEW

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.ts:42]: please use `['auth', 'getPermissions', params]` as key

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.ts:43]: This is a breaking change, as we used to pass in the params

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.ts:47]: You probably need to memo this, or this will trigger unwanted rerenders

**Review by fzaninotto — CHANGES_REQUESTED** [review]
Nice! This is definitely something we wanted to do, so it's nice of you to tackle this.

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts]: done :)

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts]: also passed now

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts:49]: Are you sure about that?
It's basically just a wrapper for `useQuery`.

Comment (fzaninotto) [reply packages/ra-core/src/auth/usePermissions.ts:49]: Yes I'm sure: `useQuery` will return an identical object if nothing has changed, but your hook will always return a new object.

Comment (fzaninotto) [reply packages/ra-core/src/auth/usePermissions.ts:49]: See #8174

File: packages/ra-core/src/auth/usePermissions.spec.tsx
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.spec.tsx:12]: please don't name the variable `authParams` as it's misleading

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.ts:37]: I'd call it `queryParams` instead of `useQueryParams`

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [inline packages/ra-core/src/auth/usePermissions.ts:41]: Since we don't need to make the effect dependent on `getPermissions` anymore, I'd use `useAuthProvider` instead of `useGetPermissions`.

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts]: done.

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts]: done, thanks!

File: packages/ra-core/src/auth/usePermissions.ts
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts]: I removed the useGetPermissions and added a conditional if authProvider is null.
does that work for you?

File: packages/ra-core/src/auth/usePermissions.spec.tsx
Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.spec.tsx]: renamed

**Comment by fzaninotto** [comment]
Nice! Waiting for a second review to merge.

File: packages/ra-core/src/auth/usePermissions.ts
Comment (djhi) [inline packages/ra-core/src/auth/usePermissions.ts:40]: Can you explain why you chose these defaults?

**Review by djhi — COMMENTED** [review]
Nice! Thanks

**Review by djhi — APPROVED** [review]

File: packages/ra-core/src/auth/usePermissions.ts
Comment (fzaninotto) [reply packages/ra-core/src/auth/usePermissions.ts:40]: Shouldn't you use a `staleTime` instead?

Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts:40]: sure, i'm fine with staleTime too.
What do you think is a sane default? 60s? 5min?

Comment (fzaninotto) [reply packages/ra-core/src/auth/usePermissions.ts:40]: 5 mins is OK

Comment (hiaselhans) [reply packages/ra-core/src/auth/usePermissions.ts:40]: updated :)

**Pull request merged**
